### PR TITLE
Exception logging to database.

### PR DESCRIPTION
**Pull request by mikegrima**
References #392

### CODE REVIEW

**Comment by mikegrima** [comment]
Almost done with PR.  Will be adding a scheduler command to clean out exceptions that are older than the TTL.

File: security_monkey/datastore.py [hunk 1]
```diff
@@ -288,6 +298,26 @@ class IgnoreListEntry(db.Model):
...
+    """
+    __tablename__ = "exceptions"
+    id = Column(BigInteger, primary_key=True)
+    source = Column(String(256), nullable=False, index=True)
+    occurred = Column(DateTime, default=datetime.datetime.utcnow(), nullable=False)
+    ttl = Column(DateTime, default=(datetime.datetime.utcnow() + datetime.timedelta(days=10)), nullable=False)
+    type = Column(String(256), nullable=False, index=True)
+    message = Column(String(512))
+    stacktrace = Column(Text)
+    region = Column(String(64), nullable=True, index=True)
```
Comment (monkeysecurity) [inline security_monkey/datastore.py]: Item.region is only 32 chars. Odd to have the same data stored with different lengths.

Comment (mikegrima) [inline security_monkey/datastore.py]: Oops.  Fixing...

File: security_monkey/decorators.py [hunk 2]
```diff
@@ -111,11 +113,12 @@ def iter_account_region(index=None, accounts=None, regions=None, exception_recor
...
         @wraps(f)
         def decorated_function(*args, **kwargs):
-            item_list = []; exception_map = {}
+            item_list = []
+            exception_map = {}
             for account_name, region in product(accounts, regions):
                 account = Account.query.filter(Account.name == account_name).first()
                 if not account:
-                    print "Couldn't find account with name",account_name
+                    print "Couldn't find account with name", account_name
```
Comment (monkeysecurity) [inline security_monkey/decorators.py]: Should this be a log statement?

File: security_monkey/scheduler.py [hunk 3]
```diff
@@ -141,6 +147,9 @@ def setup_scheduler():
             if auditors:
                 scheduler.add_cron_job(_audit_changes, hour=10, day_of_week="mon-fri", args=[account, auditors, True])
 
+                # Clear out old exceptions:
+                scheduler.add_cron_job(_clear_old_exceptions, hour=3, minute=0)
```
Comment (monkeysecurity) [inline security_monkey/scheduler.py]: It looks like you're going to schedule _clear_old_exceptions to run at 3am for every account that has an auditor.

Probably want to put this after the `for account in accounts` loop.

File: security_monkey/tests/__init__.py [hunk 4]
```diff
@@ -24,15 +24,20 @@
...
+        self.pre_test_setup()
 
+    def pre_test_setup(self):
+        # Each sub-class can implement this.
+        pass
+
+    def tearDown(self):
+        db.session.close()
+        #db.session.remove()
+        #db.drop_all()
```
Comment (monkeysecurity) [inline security_monkey/tests/__init__.py]: should we delete `#db.session.remove()` and `#db.drop_all()`?

File: security_monkey/decorators.py (hunk 2 above)
Comment (mikegrima) [inline security_monkey/decorators.py]: Fixed with: `app.logger.error("Couldn't find account with name", account_name)`

File: security_monkey/scheduler.py (hunk 3 above)
Comment (mikegrima) [inline security_monkey/scheduler.py]: Good catch.  Fixed.

File: security_monkey/tests/__init__.py (hunk 4 above)
Comment (mikegrima) [inline security_monkey/tests/__init__.py]: Done.

Comment (mikegrima) [inline security_monkey/tests/__init__.py]: Done.

File: security_monkey/watcher.py [hunk 5]
```diff
@@ -146,23 +145,26 @@ def changed(self):
...
         Location can also exclude an item_name if the exception is region wide.
         """
         if location in exception_map:
             app.logger.debug("Exception map already has location {}. This should not happen.".format(location))
         exception_map[location] = exception
         app.logger.debug("Adding {} to the exceptions list. Exception was: {}".format(location, str(exception)))
 
-    def locationInExceptionMap(self, item_location, exception_map={}):
+        # Store it to the database:
+        store_exception("watcher", location, exception)
```
Comment (monkeysecurity) [inline security_monkey/watcher.py]: The `@record_exception` decorator defaults source to "boto", but here you're setting it to "watcher".  It looks like one place `iam_role.py` sets it to "watcher", but the route53.py watcher does not set it.:   (line 41, 46, 51)

https://github.com/mikegrima/security_monkey/blob/exceptions_table/security_monkey/watchers/route53.py#L41

In `iam_role.py`, you set the source on line 74, but not on line 46:
https://github.com/mikegrima/security_monkey/blob/exceptions_table/security_monkey/watchers/iam/iam_role.py#L46

Comment (mikegrima) [inline security_monkey/watcher.py]: Fixed. 🍌

**Pull request merged**